### DISCUSSION TITLE
Keep last volume setting

### DISCUSSION THREAD

**Discussion by eki1503** [biologist79/ESPuino#71] [Q&A]
Is there a way to keep the last volume setting during reboot instead of always setting the volume to the specified startup value?

**Comment by biologist79** [comment]
Currently that's not possible. I mean, would be pretty easy to add, but does that make sense? I mean, not every track has the same volume and usually the volume used depends on your mood/situation?

**Reply by eki1503** [reply]
First of all thanks for this great project. I am using it to build a sound box for and elderly relative.
For me this is more an UI issue. The knob that changes the volume is an "endless" one. So in the current configuration, you cannot see from the position of the knob what volume currently is set. This would be different with the change (at least up to a certain extend as you can of course move the knob beyond maximum and below minimum). I agree that there are reasons for modifying the volume according to mood and track. However in today's setup you got the same problems as the volume setting is always at a default at the beginning.

**Reply by biologist79** [reply]
IMHO the advantage of the current way is': it's in a defined state every time. No surprises here.
1) Imagine you've set the volume to 21 and after reboot you'll maybe end up with "WTF!?" when the music starts to "scream".
2) Imagine because of a phone call you lowered the volume down to 0. After reboot you expect the music starting to play as you applied a RFID-tag but as it seems nothing happens. Now does your elderly relative still remember this is because the last event (maybe a couple of days ago?
Additionaly, the current volume is visible via Neopixel (as you change the volume and after bootup).

Can implement it but IMHO it's a bad idea.

**Reply by eki1503** [reply]
As I explained, you would see the setting you left to the volume by the position of the mark on the knob, so no need for anybody to remember anything (and by the way this is the behaviour my relative is used to since decades from good old transistor radios ;) ). But if you treat it as a bad idea, no problem if you do not implement it.

**Reply by biologist79** [reply]
I will add it. Had a chat with a colleague of mine to get his opinion. He found your argumentation reasonably.
Is a #define-statement sufficient? Just askin because this would overwrite an initial-volume-setting from the GUI.

**Reply by eki1503** [reply]
Great thanks. Sure, #define is ok.

**Reply by biologist79** [reply]
Just committed the feature.
Please note:
a) This will overwrite the initial volume configured via GUI (if active).
b) The volume is saved to flash when Tonuino is shutting down. This is because number of writes to flash is limited. So IMHO doesn't make sense to write that setting to flash every time the volume changes.
c) If this feature is enabled for the first time, there's no last volume. In this case initial volume from GUI will be taken. As you shutdown the Tonuino und reboot again, value saved at shutdown will be taken.

Have fun :-)

**Reply by eki1503** [reply]
Great, thanks for your fast response.

**Reply by biologist79** [reply]
Totally forgot this conversation.
This feature has been implemented a while ago and can be used by enabling USE_LAST_VOLUME_AFTER_REBOOT in settings.h.
